File: Android/app/src/main/java/antoni/kalorie/core/usecases/DeleteAccountUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthCommandProviderProtocol
import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.auth.PendingMergeSnapshotStoreProtocol
import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.networking.FavouriteFoodDTO
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodConsumedDTO
import antoni.kalorie.core.networking.FoodItemPersonalPortionsDTO
import antoni.kalorie.core.networking.FoodItemReportDTO
import antoni.kalorie.core.networking.FoodItemSubmissionDTO
import antoni.kalorie.core.networking.MealTypeDTO
import antoni.kalorie.core.networking.MyCreatedMealDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.Log
import com.google.firebase.auth.FirebaseAuthRecentLoginRequiredException
import java.time.Duration
import java.time.Instant
import kotlinx.coroutines.CancellationException

sealed class DeleteAccountError : Exception() {
    data class RequiresRecentLogin(val dataAlreadyDeleted: Boolean) : DeleteAccountError()
}

interface DeleteAccountUseCaseProtocol {
    suspend operator fun invoke(skipDataWipe: Boolean = false)
}

class DeleteAccountUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
    private val authCommandProvider: AuthCommandProviderProtocol,
    private val snapshotStore: PendingMergeSnapshotStoreProtocol,
) : DeleteAccountUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(skipDataWipe: Boolean) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val lastSignInDate = authProvider.lastSignInDate
        if (lastSignInDate != null && Duration.between(lastSignInDate, Instant.now()).seconds > Constants.Auth.RECENT_LOGIN_THRESHOLD_SECONDS) {
            throw DeleteAccountError.RequiresRecentLogin(dataAlreadyDeleted = false)
        }
        // A snapshot left by a failed merge would otherwise be resumed into the fresh anonymous
        // account that follows the deletion. Idempotent, so a retry with skipDataWipe is safe.
        snapshotStore.delete()
        if (!skipDataWipe) wipeFirestoreData(userId)
        try {
            authCommandProvider.deleteCurrentUser()
        } catch (error: CancellationException) {
            throw error
        } catch (error: FirebaseAuthRecentLoginRequiredException) {
            throw DeleteAccountError.RequiresRecentLogin(dataAlreadyDeleted = true)
        }
    }

    // MARK: - Private

    private suspend fun wipeFirestoreData(userId: String) {
        val mealTypes: List<MealTypeDTO> = dataProvider.loadAsync(from = Constants.Firestore.mealTypes(userId))
        mealTypes.forEach { dataProvider.deleteAsync(id = it.id, from = Constants.Firestore.mealTypes(userId)) }
        val foods: List<FoodConsumedDTO> = dataProvider.loadAsync(from = Constants.Firestore.foodConsumed(userId))
        foods.forEach { dataProvider.deleteAsync(id = it.id, from = Constants.Firestore.foodConsumed(userId)) }
        val favouriteFoods: List<FavouriteFoodDTO> = dataProvider.loadAsync(from = Constants.Firestore.favouriteFoods(userId))
        favouriteFoods.forEach { dataProvider.deleteAsync(id = it.id, from = Constants.Firestore.favouriteFoods(userId)) }
        val myCreatedMeals: List<MyCreatedMealDTO> = dataProvider.loadAsync(from = Constants.Firestore.myCreatedMeals(userId))
        myCreatedMeals.forEach { dataProvider.deleteAsync(id = it.id, from = Constants.Firestore.myCreatedMeals(userId)) }
        val foodItemPortions: List<FoodItemPersonalPortionsDTO> = dataProvider.loadAsync(from = Constants.Firestore.foodItemPortions(userId))
        foodItemPortions.forEach { dataProvider.deleteAsync(id = it.id, from = Constants.Firestore.foodItemPortions(userId)) }
        val submissions: List<FoodItemSubmissionDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.FOOD_ITEM_SUBMISSIONS,
            field = "submitted_by",
            isEqualTo = userId,
            orderBy = "submitted_at",
            descending = false,
        )
        submissions.forEach { dataProvider.deleteAsync(id = it.id, from = Constants.Firestore.FOOD_ITEM_SUBMISSIONS) }
        val reports: List<FoodItemReportDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.FOOD_ITEM_REPORTS,
            field = "reported_by",
            isEqualTo = userId,
            orderBy = "reported_at",
            descending = false,
        )
        reports.forEach {
            dataProvider.deleteAsync(id = FoodItemReportDomain.id(barcode = it.barcode, userId = userId), from = Constants.Firestore.FOOD_ITEM_REPORTS)
        }
        try {
            dataProvider.deleteAsync(id = userId, from = Constants.Firestore.USERS)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.ACCOUNT)
        }
    }
}

File: iOS/Kalorie/Core/UseCases/DeleteAccountUseCase.swift
//
//  DeleteAccountUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 07.08.2026.
//

import FirebaseAuth
import Foundation

enum DeleteAccountError: Error {
    case requiresRecentLogin(dataAlreadyDeleted: Bool)
}

protocol DeleteAccountUseCaseProtocol {
    func callAsFunction(skipDataWipe: Bool) async throws
}

struct DeleteAccountUseCase: DeleteAccountUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol
    private let authCommandProvider: any AuthCommandProviderProtocol
    private let snapshotStore: any PendingMergeSnapshotStoreProtocol

    // MARK: - Init

    init(
        dataProvider: any FirestoreDataProviderProtocol,
        authProvider: any AuthProviderProtocol,
        authCommandProvider: any AuthCommandProviderProtocol,
        snapshotStore: any PendingMergeSnapshotStoreProtocol
    ) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
        self.authCommandProvider = authCommandProvider
        self.snapshotStore = snapshotStore
    }

    // MARK: - Functions

    func callAsFunction(skipDataWipe: Bool = false) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }

        if
            let lastSignInDate = authProvider.lastSignInDate,
            Date().timeIntervalSince(lastSignInDate) > Constants.Auth.recentLoginThreshold
        {
            throw DeleteAccountError.requiresRecentLogin(dataAlreadyDeleted: false)
        }

        // A snapshot left by a failed merge would otherwise be resumed into the fresh anonymous
        // account that follows the deletion. Idempotent, so a retry with skipDataWipe is safe.
        try snapshotStore.delete()

        if !skipDataWipe {
            try await wipeFirestoreData(userId: userId)
        }

        do {
            try await authCommandProvider.deleteCurrentUser()
        } catch {
            guard error.matches(domain: AuthErrorDomain, code: AuthErrorCode.requiresRecentLogin.rawValue) else { throw error }
            throw DeleteAccountError.requiresRecentLogin(dataAlreadyDeleted: true)
        }
    }

    private func wipeFirestoreData(userId: String) async throws {
        let mealTypes: [MealTypeDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.mealTypes(userId: userId))
        for dto in mealTypes {
            try await dataProvider.deleteAsync(id: dto.id, from: Constants.Firestore.mealTypes(userId: userId))
        }

        let foods: [FoodConsumedDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.foodConsumed(userId: userId))
        for dto in foods {
            try await dataProvider.deleteAsync(id: dto.id, from: Constants.Firestore.foodConsumed(userId: userId))
        }

        let favouriteFoods: [FavouriteFoodDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.favouriteFoods(userId: userId))
        for dto in favouriteFoods {
            try await dataProvider.deleteAsync(id: dto.id, from: Constants.Firestore.favouriteFoods(userId: userId))
        }

        let myCreatedMeals: [MyCreatedMealDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.myCreatedMeals(userId: userId))
        for dto in myCreatedMeals {
            try await dataProvider.deleteAsync(id: dto.id, from: Constants.Firestore.myCreatedMeals(userId: userId))
        }

        let foodItemPortions: [FoodItemPersonalPortionsDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.foodItemPortions(userId: userId))
        for dto in foodItemPortions {
            try await dataProvider.deleteAsync(id: dto.id, from: Constants.Firestore.foodItemPortions(userId: userId))
        }

        let submissions: [FoodItemSubmissionDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.foodItemSubmissions,
            where: "submitted_by",
            isEqualTo: userId,
            orderBy: "submitted_at",
            descending: false
        )
        for dto in submissions {
            try await dataProvider.deleteAsync(id: dto.id, from: Constants.Firestore.foodItemSubmissions)
        }

        let reports: [FoodItemReportDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.foodItemReports,
            where: "reported_by",
            isEqualTo: userId,
            orderBy: "reported_at",
            descending: false
        )
        for dto in reports {
            try await dataProvider.deleteAsync(
                id: FoodItemReportDomain.id(barcode: dto.barcode, userId: userId),
                from: Constants.Firestore.foodItemReports
            )
        }

        do {
            try await dataProvider.deleteAsync(id: userId, from: Constants.Firestore.users)
        } catch {
            Log.error(error, category: Constants.LogCategory.account)
        }
    }
}

#if DEBUG
struct DeleteAccountUseCaseFake: DeleteAccountUseCaseProtocol {

    // MARK: - Properties

    var errorToThrow: Error?

    // MARK: - Functions

    func callAsFunction(skipDataWipe: Bool = false) async throws {
        if let errorToThrow { throw errorToThrow }
    }
}
#endif
